KAFKA-14668 Avoid unnecessary UMR during ZK migration

mumrah opened the pull request:
Prior to this patch, when we handle a new MetadataDelta during dual write mode in the KRaft controller, we always send UMR to the ZK brokers. This patch changes the behavior to only send UMR if the broker or topic metadata has changed (since this is all that we send via UMR).

A unit test for KRaftMigrationDriver is added to validate this behavior. 

Some improved error handling and logging in KRaftMigrationDriver is also included.

mumrah committed:
Fix the bug and add unit test

mumrah committed:
style cleanup

cmccabe commented on metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java:
```diff
@@ -106,6 +109,13 @@ public void shutdown() throws InterruptedException {
     }
 
+    // Visible for testing
+    Future<MigrationState> migrationState() {
```
Would be better as if the return type were `CompletableFuture`

cmccabe commented on metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java:
```diff
@@ -474,6 +472,7 @@ class MetadataChangeEvent extends MigrationEvent {
         private final MetadataImage image;
         private final MetadataProvenance provenance;
         private final boolean isSnapshot;
+        CompletableFuture<Void> future = new CompletableFuture<>();
```
do we need this? it's expensive, and I can find 0 uses of it

Hangleton replied:
If we want to follow this pattern where the tests `KRaftMigrationDriverTest` are signalled when this event is processed, and at the same time avoid the cost of adding a future to every `SendRPCsToBrokersEvent`, `enqueueMetadataChangeEvent` could take a future in argument instead. Prod code would pass an empty/noop future. Alternatively the unit tests could loop until the assertions are verified or a timeout is reached.

mumrah replied:
It's used to avoid having a bunch of wait-for conditions in the tests. Let me try your suggestion @Hangleton

cmccabe replied:
Thanks, looks good now.

cmccabe commented on server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java:
```diff
@@ -250,7 +250,7 @@ private void handleEvents() throws InterruptedException {
                         } else if (shuttingDown) {
                             remove(eventContext);
-                            toDeliver = new TimeoutException();
+                            toDeliver = new QueueClosingException();
```
Hmm, I'm not really sure about having a new execption here. Can we just use `RejectedExecutionException` here?

cmccabe replied:
Also we will need a test in `KafkaEventQueueTest` if you want to do this. I do wonder if this should be a separate PR ... queue changes have a way of being big. But I guess let's see if it impacts any other tests.

mumrah replied:
I thought about `RejectedExecutionException`, but wasn't sure if we relied on the `TimeoutException` type somewhere, so I made a subclass of that. If you think it's fine, we'll go with `RejectedExecutionException`.

cmccabe replied:
I think `RejectedExecutionException` really is more appropriate. If you tried to schedule (or reschedule) a deferred task and the queue was shut down, you'd get `RejectedExecutionException` already. So you already have to catch REE and treat it this way.

mumrah committed:
Feedback from PR

mumrah committed:
Use RejectedExectionException instead of QueueClosingException

mumrah commented:
Updates:
* Factored out the CompletableFuture from the MetadataChangeEvent so it's only used in the unit test
* Removed QueueClosingException, replaced with RejectedExecutionException, updated unit test

cmccabe commented on metadata/src/test/java/org/apache/kafka/metadata/migration/KRaftMigrationDriverTest.java:
```diff
@@ -0,0 +1,315 @@
+
+    @Test
+    public void testOnlySendNeededRPCsToBrokers() throws Exception {
+        // KAFKA-14668 Don't send RPCs to brokers for every metadata change, only when broker or topics change.
```
Maybe a JavaDoc that explains "This is a regression test for KAFKA-14668..."

cmccabe commented on metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java:
```diff
@@ -538,12 +553,23 @@ public void run() throws Exception {
+                    propagator.sendRPCsToBrokersFromMetadataDelta(delta, image,
+                            migrationLeadershipState.zkControllerEpoch());
+                } else {
+                    log.trace("Not sending RPCs to brokers for metadata {} since no relevant metadata changes", metadataType);
```
nitpick: perhaps "since no relevant metadata has changd"

cmccabe commented on metadata/src/test/java/org/apache/kafka/metadata/migration/KRaftMigrationDriverTest.java:
```diff
@@ -0,0 +1,315 @@
+            .setValue("bar"));
+        provenance = new MetadataProvenance(120, 1, 2);
+        image = delta.apply(provenance);
+        enqueueMetadataChangeEventWithFuture(driver, delta, image, provenance).get(10, TimeUnit.SECONDS);
```
maybe 1 minute as a timeout? The Jenkins environment can get wacky so short timeouts are almost never a good idea.

mumrah committed:
More PR feedback, fix regression in another test

FireBurn commented:
Has ZK migration to Kraft been delayed?

mumrah commented:
@FireBurn, no it hasn't been delayed. We have an early access version coming out in 3.4.0 which will be announced this week. We plan on back-porting the remaining ZK migration work to the 3.4.x line so we can continue releasing without waiting for the next major release.

FireBurn commented:
Fab, I've just built 3.4.0 from the tag, but couldn't find the migration docs, so worried it had been delayed to 3.4.1

mumrah commented:
Gotcha, yea the docs are still in progress. We'll have something published soon (hopefully before the release announcement)

akhileshchg commented on metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java:
```diff
@@ -274,11 +310,6 @@ public void run() throws Exception {
                 new EventQueue.DeadlineFunction(deadline),
                 new PollEvent());
         }
-
```
Why are we not handling exceptions in other events?

mumrah replied:
The base MigrationEvent will do the error logging. I was just removing some redundant overrides.

cmccabe reviewed (approved):
LGTM
